### ISSUE TITLE
(I)Python tab completion

### ISSUE THREAD

**Issue by gf712** [shogun-toolbox/shogun#4476]
The default Python shell can be extended with tab completion using pyreadline and IPython uses it's own implementation now for tab completion. (Tab completion refers to completing an object name with `<TAB>`, so for example when you `import shogun as sg` you can do `sg.ma<press TAB>`  and you would probably get `sg.machine`).
It would be useful to have something like that for all the strings that correspond to the names of all shogun objects.
Currently I thought about something like this for IPython (>=5.0, the version where IPython removed pyreadline):
```python
from IPython.core.completerlib import quick_completer
import shogun as sg
all_sg_objects = sg.available_objects() # needs to be exposed via SWIG
quick_completer("\"", all_sg_objects) # tab completion for open quotes (")
``` 
This adds tab completion when a user starts typing a Python string (`"string"`). The issue is that this seems to completely override the default behaviour. So typing `"` will now only show tab completion with shogun objects. However if you start typing the remaining object names appear again.

**Comment by AnirudhDagar** [comment]
Looks like an interesting problem to solve.
It would be nice if you could explain the last three lines of the issue. What do you mean by overriding default behavior specifically?
Also, any pointers to start with would be really helpful as this will be my first issue.

**Comment by gf712** [comment]
Hey, sure I can help out. I have a somewhat stable solution, but it's not very elegant. Also need @karlnapf's input, as I am not sure if this is a direction we want to go in (or at least right now).

Sorry, it's a bit hard to explain. When you add completions in IPython with quick_completer you tell it that from that point on only those completions should be shown after the given pattern. Before calling quick_completer however when you write that pattern there are other completions suggested (so these disappear, that's what I mean with "The issue is that this seems to completely override the default behaviour".

Anyway, this is what I got so far:
```python
import shogun as sg 
from IPython.core import getipython

def load_shogun_objects(regex_pattern, completions):
	"""
	IPython completer
	"""
	def do_complete(self, event):
		return completions

	get_ipython().set_hook('complete_command', do_complete, re_key=regex_pattern)

load_shogun_objects("[\w|\.]+\((name\s*=){0,1}\s*\"\w*", sg._available_objects())
```
Which means this can be autocompleted:
```python
sg.kernel(name="G<TAB>
```
to:
```python
sg.kernel(name="Gaussian<TAB>
```
and then 
```python
sg.kernel(name="GaussianKernel`
```
If you are curious as to how to get sg._available_objects check #4473. And if that looks interesting you could get started with #4475, which helps getting exposure with how the C++ library works and how to expose it via SWIG (and is already labelled as a good first issue)

**Comment by karlnapf** [comment]
Completely overriding the default behaviour is tricky indeed.
In the perfect world we would have `sg.kernel("G...` only suggest from the kernel names. This would require some kind of nested namespacing though. But OK, let's maybe start with all shogun objects for now. There are other way to achieve this, e.g. via defining (string) constants with the `sg` module
```sg.kernel(sg.Gaus<TAB>```

@lisitsyn you had some ideas on this recently ...any thoughts?

I think in general we need to play around to see whether we can arrive at something elegant @gf712 @AnirudhDagar 

Great initiative this is though! :)

**Comment by karlnapf** [comment]
Well anything that works nicely is more than nothing :)

**Comment by gf712** [comment]
True, I personally really like the tab completion, it has already helped me a bit when testing out shogun python code, but I am also biased because I wrote that snippet :)

**Comment by karlnapf** [comment]
Me too. I am all for adding this and then improve it, go with the flow. Python changes all the time anyways and some smart kid will come along with a better solution in the future I am sure ... :)

**Comment by gf712** [comment]
@AnirudhDagar do you want to have a go at it? Maybe you can come up with something better? There is also the Jedi python package which is now used in IPython, and might work better, but I have never used it.

**Comment by matloob1191** [comment]
We load, haul, and dispose of your junk while offering the LOWEST POSSIBLE junk removal RATES and service around. kindly visit www.dubaijunkremoval.com

**Comment by karlnapf** [comment]
Reported as abuse with github
